src/compiler/checker.ts

```diff
@@ -5357,7 +5378,12 @@ namespace ts {
         }
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            const areAllParametersUntyped = !forEach(node.parameters, p => p.type);
+            if (node.kind === SyntaxKind.ArrowFunction) {
+                return !node.typeParameters && node.parameters.length && areAllParametersUntyped;
+            }
+            const hasThisType = node.parameters.length && (<Identifier>node.parameters[0].name).text === "this" && node.parameters[0].type;

```

Comment:
this check is subsumed by the above check for areAllParametersUntyped. do we need it?